datumaro/plugins/open_images_format.py

```diff
@@ -0,0 +1,155 @@
+)
+
+# image IDs in Open Images are always hexadecimal numbers
+_RE_IMAGE_ID = re.compile(r'[0-9a-fA-F]+')
+
+class _OpenImagesPath:
+    ANNOTATIONS_DIR = 'annotations'
+    FULL_IMAGE_DESCRIPTION_NAME = 'image_ids_and_rotation.csv'
+    SUBSET_IMAGE_DESCRIPTION_PATTERNS = (
+        '*-images-with-rotation.csv',
+        '*-images-with-labels-with-rotation.csv',
+    )
+
+class OpenImagesExtractor(Extractor):
+    def __init__(self, path):
+        if not osp.isdir(path):
+            raise FileNotFoundError("Can't read dataset directory '%s'" % path)
+
+        super().__init__()
+
+        self._dataset_dir = path
+        self._categories = {}
+        self._items = []
+
+        self._load_categories()
+        self._load_items()
+
+    def __iter__(self):
+        return iter(self._items)
+
+    def categories(self):
+        return self._categories
+
+    @contextlib.contextmanager
+    def _open_csv_annotation(self, file_name):
+        absolute_path = osp.join(self._dataset_dir, _OpenImagesPath.ANNOTATIONS_DIR, file_name)
+
+        with open(absolute_path, 'r', newline='') as f:
+            yield csv.DictReader(f)
+
+    def _glob_annotations(self, pattern):
+        return glob.glob(osp.join(
+            glob.escape(self._dataset_dir), _OpenImagesPath.ANNOTATIONS_DIR, pattern))
+
+    def _load_categories(self):
+        label_categories = LabelCategories()
+
+        with self._open_csv_annotation('oidv6-class-descriptions.csv') as class_description_reader:
+            for class_description in class_description_reader:
+                label_categories.add(class_description['LabelName'])
+
+        self._categories[AnnotationType.label] = label_categories
+
+        self._load_label_category_parents()
+
+    def _load_label_category_parents(self):
+        label_categories = self._categories[AnnotationType.label]
+
+        hierarchy_path = osp.join(
+            self._dataset_dir, _OpenImagesPath.ANNOTATIONS_DIR, 'bbox_labels_600_hierarchy.json')
+
+        with open(hierarchy_path, 'rb') as hierarchy_file:
+            root_node = json.load(hierarchy_file)
+
+        def set_parents_from_node(node, category):
+            for child_node in node.get('Subcategory', []):
+                _, child_category = label_categories.find(child_node['LabelName'])
+
+                if category is not None and child_category is not None:
+                    child_category.parent = category.name
+
+                set_parents_from_node(child_node, child_category)
+
+        _, root_category = label_categories.find(root_node['LabelName'])
+        set_parents_from_node(root_node, root_category)
+
+    def _load_items(self):
+        items_by_id = {}
+
+        def load_from(path):
+            with self._open_csv_annotation(path) as image_reader:
+                for image_description in image_reader:
+                    image_id = image_description['ImageID']
+                    assert _RE_IMAGE_ID.fullmatch(image_id)
```

Comment:
The situation with image IDs is the same as with subset names - they are used in file names, so some sort of validation is required. I can loosen the restrictions, though, as the only real constraint is that the IDs must be filename-safe.

Reply:
We don't restrict image ids in having slashes, because they can be a relative paths. The only exception I can imagine - we need to check for a leading slash, because this would define an absolute path. I don't think we must restrict them in any else way, because formats can have different ways to store data.

Reply:
I think we would also need to prohibit `..` path components, since that would allow accessing files outside the dataset.

Reply:
I loosened up the subset name validation rules (and replaced the assert with a custom exception).

Reply:
How about adding `id` and `subset` name checks into the core codebase? Having a check for abspath and for leading dots looks more or less generally-applicable. As well as for subset name being a valid filename / dirname.

Reply:
It's a good idea, but can we maybe defer it until after this PR? It sounds like a lengthy task in itself, since I would need to make sure that these checks don't interfere with loading the currently-supported datasets.

In addition, I'm not so sure that we can use the same check for every format. Some rules can probably be made universal (e.g. no empty string, no control characters), but some datasets might be okay with, for example, slashes in IDs. (There is sort of a precedent with Open Images itself, since it has _label_ names that contain slashes).

Reply:
I'd vote for another independent PR with this functionality. Slashes in ids are ok, because ids are relative paths inside dataset by design. Everything else - probably, not ok.

Reply:
Created an issue for this: #313.